### PR TITLE
migrate registries tests

### PR DESCRIPTION
**Pull request by gfournieriExec**
Coverage should better Just can't test one branch with the re - initialisation of ressources 

image
image

### CODE REVIEW

File: test/byContract/registries/registries.test.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    beforeEach(async () => {
+        proxyAddress = await loadHardhatFixtureDeployment();
+        await loadFixture(initFixture);
+    });
+
+    async function initFixture() {
+        ({ iexecAdmin, appProvider, datasetProvider, scheduler, anyone } =
+            await getIexecAccounts());
+        ensRegistryAddress = (await deployments.get('ENSRegistry')).address;
+        const appRegistryAddress = (await deployments.get('AppRegistry')).address;
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:43]: I propose to replace with `0xpoco.appregistry()` to remove some adhesion

File: test/byContract/registries/registries.test.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        appRegistry = AppRegistry__factory.connect(appRegistryAddress, iexecAdmin);
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:44]: connect with anyone, then later with specific admin account for transparency

File: test/byContract/registries/registries.test.ts [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+            expect(await appRegistry.baseURI()).to.equal(
+                `https://nfts-metadata.iex.ec/app/${chainId}/`,
+            );
+            expect(await datasetRegistry.baseURI()).to.equal(
+                `https://nfts-metadata.iex.ec/dataset/${chainId}/`,
+            );
+            expect(await workerpoolRegistry.baseURI()).to.equal(
+                `https://nfts-metadata.iex.ec/workerpool/${chainId}/`,
+            );
+        });
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:69]: Hard to add should set name ?

File: test/byContract/registries/registries.test.ts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                appRegistry.connect(anyone).setBaseURI(`https://new.url.iex.ec/app/`),
+            ).to.be.revertedWith('Ownable: caller is not the owner');
+            await expect(
+                datasetRegistry.connect(anyone).setBaseURI(`https://new.url.iex.ec/dataset/`),
+            ).to.be.revertedWith('Ownable: caller is not the owner');
+            await expect(
+                workerpoolRegistry.connect(anyone).setBaseURI(`https://new.url.iex.ec/workerpool/`),
+            ).to.be.revertedWith('Ownable: caller is not the owner');
+        });
+        it('Should not call initialize when user is not the owner', async () => {
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:96]: Why not adding should initialize with fresh registries ?

File: test/byContract/registries/registries.test.ts [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        workerpoolRegistry = WorkerpoolRegistry__factory.connect(
+            workerpoolRegistryAddress,
+            iexecAdmin,
+        );
+
+        iexecPoco = IexecInterfaceNative__factory.connect(proxyAddress, anyone);
+        iexecPocoAsAdmin = iexecPoco.connect(iexecAdmin);
+    }
+
+    describe('Registry', () => {
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:57]: Remove this describe then 1 describe per function?

File: test/byContract/registries/registries.test.ts (hunk 2 above)
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts]: Done update [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/ebf64ef8546e0415cbc666d4c53fef26b5bffec1)

Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts]: Updated [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/8123e78ca8a9c2a22e24c98b1f3023345bd759ac)

File: test/byContract/registries/registries.test.ts (hunk 4 above)
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:121]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/f3ac7c22e9d7f41aee95143c75578fdc7d17ae04)

File: test/byContract/registries/registries.test.ts (hunk 3 above)
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:138]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/a180f4b057dfdc54d2bb3c8187c23823bfc2b685)

File: test/byContract/registries/registries.test.ts (hunk 5 above)
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/812e972a8d71789673e739a09b8f77f66648a42c) tell me if that's okay for you

File: deploy/0_deploy.ts [hunk 6]
```diff
@@ -146,6 +146,43 @@ module.exports = async function () {
...
+        owner,
+    );
+
+    // Check if registries have been initialized and set base URIs
+    const appRegistryInitialized = await appRegistryInstance.initialized();
+    if (!appRegistryInitialized) {
+        await appRegistryInstance.initialize(
+            deploymentOptions.v3.AppRegistry || ethers.constants.AddressZero,
+        );
+        await appRegistryInstance.setBaseURI(`https://nfts-metadata.iex.ec/app/${chainId}/`);
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply deploy/0_deploy.ts:163]: You want me to parameterized it in config.json for exemple ?

[the remark this answers is not in the document]
Comment (zguesmi) [reply deploy/0_deploy.ts:163]: It makes sense, yes.

File: test/byContract/registries/registries.test.ts [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+            '0x1234',
+        ] as [string, string, BytesLike, BytesLike, BytesLike];
+        it('Should predict the correct address for future app creation', async () => {
+            const code = await appRegistry.proxyCode();
+            const proxyCodeHash = ethers.utils.keccak256(code);
+
+            const iface = new ethers.utils.Interface([
+                'function initialize(string,string,bytes,bytes32,bytes)',
+            ]);
+            const encodedInitializer = iface.encodeFunctionData('initialize', createAppArgs);
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:204]:
```
            const encodedInitializer = App__factory.createInterface().encodeFunctionData(
                'initialize',
                createAppArgs,
            );
```
or
```
            const encodedInitializer = await new App__factory()
                .attach(AddressZero) // any is fine
                .populateTransaction.initialize(...createAppArgs)
                .then((populatedTransaction) => populatedTransaction.data);
```

File: test/byContract/registries/registries.test.ts [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+            ).to.be.revertedWith('Ownable: caller is not the owner');
+        });
+    });
+
+    describe('App Registry', () => {
+        const createAppArgs = [
+            `App`,
+            'DOCKER',
+            constants.MULTIADDR_BYTES,
+            ethers.utils.keccak256(ethers.utils.toUtf8Bytes(`Content of my app`)),
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:188]: it works thanks

File: test/byContract/registries/registries.test.ts [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+            ).to.be.revertedWithoutReason();
+            await expect(
+                workerpoolRegistryAsAdmin.initialize(AddressZero),
+            ).to.be.revertedWithoutReason();
+        });
+    });
+
+    describe('setName', () => {
+        it('should set the ENS name for registries', async () => {
+            const appRegistryENSName = 'myAppRegistry.eth';
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:116]: Minor, you can update all you `somethingENSVar` to `somethingEnsVar` to follow practice we use so far

File: test/byContract/registries/resources.test.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
+// SPDX-FileCopyrightText: 2020-2024 IEXEC BLOCKCHAIN TECH <contact@iex.ec>
```
Comment (jeremyjams) [inline test/byContract/registries/resources.test.ts:1]: I don't see `isRegistered` test (?)

File: test/byContract/registries/resources.test.ts [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+// SPDX-License-Identifier: Apache-2.0
```
Comment (jeremyjams) [inline test/byContract/registries/resources.test.ts:2]: Not sure to have seen "should not initialize asset (resource) twice" (?)

File: test/byContract/registries/resources.test.ts [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+            ethers.utils.keccak256(ethers.utils.toUtf8Bytes(`Content of my app`)),
+            '0x1234',
+        ] as [string, string, BytesLike, BytesLike, BytesLike];
+
+        beforeEach(async () => {
+            const appAddress = await appRegistry.callStatic.createApp(
+                appProvider.address,
+                ...createAppArgs,
+            );
+            await appRegistry.createApp(appProvider.address, ...createAppArgs);
```
Comment (jeremyjams) [inline test/byContract/registries/resources.test.ts:85]: then tx.wait
(please update everywhere!)

File: test/byContract/registries/resources.test.ts [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            app = App__factory.connect(appAddress, appProvider);
```
Comment (jeremyjams) [inline test/byContract/registries/resources.test.ts:86]: Load app with anyone, then explicit .`connect(appProvider)` elsewhere

File: test/byContract/registries/resources.test.ts [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+            expect(await workerpool.owner()).to.equal(scheduler.address);
+            expect(await workerpool.m_workerpoolDescription()).to.equal(createWorkerpoolArgs[0]);
+            expect(await workerpool.m_workerStakeRatioPolicy()).to.equal(30);
+            expect(await workerpool.m_schedulerRewardRatioPolicy()).to.equal(1);
+        });
+
+        it('should allow the owner to configure the workerpool', async () => {
+            await workerpool.changePolicy(35, 5, { from: scheduler.address });
+            expect(await workerpool.m_workerStakeRatioPolicy()).to.equal(35);
+            expect(await workerpool.m_schedulerRewardRatioPolicy()).to.equal(5);
```
Comment (jeremyjams) [inline test/byContract/registries/resources.test.ts:164]: Check event

File: test/byContract/registries/resources.test.ts (hunk 10 above)
Comment (gfournieriExec) [reply test/byContract/registries/resources.test.ts]: Will do it later

File: test/byContract/registries/resources.test.ts (hunk 12 above)
Comment (gfournieriExec) [reply test/byContract/registries/resources.test.ts]: Will do it later

File: test/byContract/registries/resources.test.ts (hunk 13 above)
Comment (gfournieriExec) [reply test/byContract/registries/resources.test.ts]: Will do it later

File: test/byContract/registries/resources.test.ts (hunk 14 above)
Comment (gfournieriExec) [reply test/byContract/registries/resources.test.ts]: Will do it later

File: test/byContract/registries/resources.test.ts [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        });
+
+        it('should create an app and verify its details', async () => {
+            expect(await app.registry()).to.equal(appRegistry.address);
+            expect(await app.owner()).to.equal(appProvider.address);
+            expect(await app.m_appName()).to.equal(createAppArgs[0]);
+            expect(await app.m_appType()).to.equal(createAppArgs[1]);
+            expect(await app.m_appMultiaddr()).to.equal(createAppArgs[2]);
+            expect(await app.m_appChecksum()).to.equal(createAppArgs[3]);
+            expect(await app.m_appMREnclave()).to.equal(createAppArgs[4]);
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply test/byContract/registries/resources.test.ts]: Will do it later

File: test/byContract/registries/registries.test.ts [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+                .connect(iexecAdmin)
+                .deploy()
+                .then((contract) => contract.deployed());
+            await newWorkerpoolRegistry
+                .initialize(workerpoolRegistry.address)
+                .then((tx) => tx.wait());
+            expect(await newWorkerpoolRegistry.initialized()).to.be.true;
+            expect(await newWorkerpoolRegistry.previous()).to.equal(workerpoolRegistry.address);
+        });
+        it('Should not call initialize when user is not the owner', async () => {
```
Comment (zguesmi) [inline test/byContract/registries/registries.test.ts:95]:
```suggestion
        });

        it('Should not call initialize when user is not the owner', async () => {
```

File: config/config.json [hunk 17]
```diff
@@ -181,5 +181,10 @@
 				"usefactory": true
 			}
 		}
+	},
+	"baseURI": {
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply config/config.json]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/a89b38fca586c11ed063dd28996ed175d357b5a8)

File: test/byContract/registries/registries.test.ts (hunk 16 above)
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/deb15ad27404937bf8baf44522c16848bf612eb1)

File: test/byContract/registries/registries.test.ts [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+        workerpoolRegistryAsAdmin = workerpoolRegistry.connect(iexecAdmin);
+    }
+
+    describe('initialize', () => {
+        it('Should initialize new deployed registries', async () => {
+            const newAppRegistry = await new AppRegistry__factory()
+                .connect(iexecAdmin)
+                .deploy()
+                .then((contract) => contract.deployed());
+            await newAppRegistry.initialize(appRegistry.address);
```
[the remark this answers is not in the document]
Comment (jeremyjams) [reply test/byContract/registries/registries.test.ts]: Great thanks

File: test/byContract/registries/registries.test.ts [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+            const predictedAddress = await appRegistry.predictApp(
+                appProvider.address,
+                ...createAppArgs,
+            );
+            await expect(await appRegistry.createApp(appProvider.address, ...createAppArgs))
+                .to.emit(appRegistry, 'Transfer')
+                .withArgs(
+                    AddressZero,
+                    appProvider.address,
+                    ethers.BigNumber.from(predictedAddress).toString(),
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:255]: I think here is a good place to check, don't you?
```
        _holderTokens[to].add(tokenId);
        _tokenOwners.set(tokenId, to);
```

Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:255]: is this not testing the ERC721 standard ? I can add it yes

Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:258]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/122/commits/340735c645cc4a8c9f311561a163d068810b9174)

Comment (jeremyjams) [reply test/byContract/registries/registries.test.ts:258]: IMO we should test when possible all behaviors of a contract, for instance we need to check that _mintCreate/_mint are called on createApp. I hope you & @zguesmi agree (?)
What you added is good to me :+1:
Though if I'm not wrong checks are missing for dataset & workerpool (?)

Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts:258]: On my way!

File: config/config.json [hunk 20]
```diff
@@ -181,5 +181,10 @@
 				"usefactory": true
 			}
 		}
+	},
+	"registriesBaseUri": {
+		"app": "https://nfts-metadata.iex.ec/app/",
```
Comment (jeremyjams) [inline config/config.json:186]:
```suggestion
		"app": "https://nfts-metadata.iex.ec/app",
```
and add slash below

File: test/byContract/registries/registries.test.ts [hunk 21]
```diff
@@ -0,0 +1,475 @@
...
+    describe('App Registry', () => {
+        const createAppArgs = [
+            `App`,
+            'DOCKER',
+            constants.MULTIADDR_BYTES,
+            ethers.utils.id(`Content of my app`),
+            '0x1234',
+        ] as [string, string, BytesLike, BytesLike, BytesLike];
+        it('Should predict the correct address for future app creation', async () => {
+            const code = await appRegistry.proxyCode();
```
Comment (jeremyjams) [inline test/byContract/registries/registries.test.ts:224]: I think coverage getters are missing for 
```
	address   public master;
	bytes     public proxyCode;
	bytes32   public proxyCodeHash;
	IRegistry public previous;
	bool      public initialized;
```
I primarily think about `Should get proxy code hash` where we should check `const proxyCodeHash = ethers.utils.keccak256(code);`

What do think guys (@zguesmi too)?

File: config/config.json (hunk 20 above)
Comment (gfournieriExec) [reply config/config.json]: updated

File: deploy/0_deploy.ts [hunk 22]
```diff
@@ -146,6 +146,42 @@ module.exports = async function () {
...
     );
+
+    const appRegistryInstance = AppRegistry__factory.connect(appRegistryAddress, owner);
+    const datasetRegistryInstance = DatasetRegistry__factory.connect(datasetRegistryAddress, owner);
+    const workerpoolRegistryInstance = WorkerpoolRegistry__factory.connect(
+        workerpoolRegistryAddress,
+        owner,
+    );
+    // Base URI configuration from config.json
+    const baseURIApp = CONFIG.registriesBaseUri.app || 'https://default.app.url/';
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply deploy/0_deploy.ts]: updated

File: test/byContract/registries/registries.test.ts [hunk 23]
```diff
@@ -0,0 +1,475 @@
...
+            expect(await reverseResolver.name(appRegistryNameHash)).to.equal('apps.v5.iexec.eth');
+            expect(await reverseResolver.name(datasetRegistryNameHash)).to.equal(
+                'datasets.v5.iexec.eth',
+            );
+            expect(await reverseResolver.name(workerpoolRegistryNameHash)).to.equal(
+                'workerpools.v5.iexec.eth',
+            );
+        });
+
+        it('should set the ENS name for registries', async () => {
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply test/byContract/registries/registries.test.ts]: updated

**Pull request merged**